firmware/python_modules/sha2017/buttons.py

```diff
@@ -1,165 +1,175 @@

-	mpr121.attach    ( __mprMap[BTN_RIGHT],  __cb_btn_right  ) # The RIGHT button is connected to input 5 of the MPR121

-	mpr121.attach    ( __mprMap[BTN_UP],     __cb_btn_up     ) # The UP button is connected to input 6 of the MPR121

-	mpr121.attach    ( __mprMap[BTN_LEFT],   __cb_btn_left   ) # The LEFT button is connected to input 7 of the MPR121

-	pushMapping() #Add the initial / default mapping

-	

-# --- PUBLIC API ---

-

-def attach(button, callback):

-	# This function attachs a callback to a button

-	global __num, __cb

-	if button < 0 or button >= __num:

-		raise ValueError("Invalid button!")

-	__cb[-1][button] = callback

-

-def detach(button):

-	# This function removes the callback of a button

-	global __num, __cb

-	if button < 0 or button >= __num:

-		raise ValueError("Invalid button!")

-	__cb[-1][button] = None

-

-def value(button):

-	# Reads the state of a button

-	global _buttons, __mprMap, __num

-	if button < 0 or button >= __num:

-		raise ValueError("Invalid button!")

-	if button == BTN_FLASH:

-		return not _buttons.pin(0).value() # This input has is active LOW

-	else:

-		return mpr121.get(__mprMap[button])

-

-def getCallback(button):

-	# Returns the currently attached callback function

-	global __num, __cb

-	if button < 0 or button >= __num:

-		raise ValueError("Invalid button!")

-	return __cb[-1][button]

-

-def pushMapping(newMapping=None):

-	global __cb

-	if newMapping == None:

-		newMapping = { BTN_UP: None, BTN_DOWN: None, BTN_LEFT: None, BTN_RIGHT: None, BTN_A: None, BTN_B: None, BTN_SELECT: None, BTN_START: None }

-		if machine.nvs_getint("system", 'factory_checked'):

-			newMapping[BTN_START] = __cbReboot

-	__cb.append(newMapping)

-

-def popMapping():

-	global __cb

-	if len(__cb) > 0:

-		__cb = __cb[:-1]

-	if len(__cb) < 1:

-		pushMapping()

-

-def rotate(value):

-	global __orientation

-	__orientation = value

-

-# ---

-__init()

+# SHA2017, Hackerhotel 2019 and Disobey 2020 button input wrapper
+# Versions of this file tailored for other badges should expose the same API
+
+import mpr121, _buttons, _mpr121mapping, system, machine
+
+# --- BUTTON CONSTANTS  ---
+BTN_A      = 0
+BTN_B      = 1
+BTN_START  = 2
+BTN_SELECT = 3
+BTN_DOWN   = 4
+BTN_RIGHT  = 5
+BTN_UP     = 6
+BTN_LEFT   = 7
+BTN_FLASH  = 8
+
+# Internal variables
+__num = 9
+__mprMap = _mpr121mapping.buttons
+__orientation = 0
+__cb = []
+
+# Internal functions
+def __cbReboot(pressed):
+    if pressed:
+        system.launcher()
+
+def __cb_btn_a(arg):
+    if __cb[-1][BTN_A]:
+        __cb[-1][BTN_A](arg)
+
+def __cb_btn_b(arg):
+    if __cb[-1][BTN_B]:
+        __cb[-1][BTN_B](arg)
+
+def __cb_btn_start(arg):
+    if __cb[-1][BTN_START]:
+        __cb[-1][BTN_START](arg)
+
+def __cb_btn_select(arg):
+    if __cb[-1][BTN_SELECT]:
+        __cb[-1][BTN_SELECT](arg)
+
+def __cb_btn_down(arg):
+    btn = BTN_DOWN

```

Comment:
This can probably be refactored into one function so we don't have to copy it four times

Reply:
Could be, but for now it works and these functions have not changed in this PR. Please create a separate ticket if you care enough :-)